Allow remote service creds to be configured using env

brandonc opened the pull request:
# Description

Introduces a new method of configuring token service credentials using a host-specific environment variable. This configuration was previously possible using the [terraform-credentials-env](https://github.com/apparentlymart/terraform-credentials-env) credentials helper.

This new method is now consulted first, as it is seen to be the most proximate source of credentials before CLI configuration. The configured credential helpers are consulted as a last resort.

The same naming methodology found in the helper is used for domains: periods (.) must be replaced by underscores, which is preferred because underscores are not allowed in DNS names and because periods are not allowed (or tricky to set) in variable names. The only gotcha I uncovered was that dashes are not valid characters in bash identifiers, while they are used in both DNS and the ACE prefix for internationalized names. Go will happily fetch a variable named "TF_TOKEN_xa--r8j3dr99h_com" but it's not possible to set this variable name using Bash without help from the "env" utility (see below)

# Testing Steps

0. Build the app: `go build -o terraform-dev`
1. Ensure you DON'T have proper credentials for your chosen test environment hostname specified in your CLI config by temporarily changing them. Make sure you keep a copy of the correct token `~/.terraform.d/credentials.tfrc.json`
2. Try to init any config with a terraform cloud block. You should receive an error because the credentials are not correct. See Sample Config below
3. Use the correct token to configure the same init: `env 'TF_TOKEN_app_terraform_io="CORRECT_TOKEN"' ./terraform-dev init`
4. If it works, this proves that the env value was used and took precedence over the credentials block.

<details>
<summary>Sample config</summary>

```
terraform {
  cloud {
    hostname = "app.terraform.io"
    organization = "YOUR_ORG"

    workspaces {
      name = "terraform-minimal-FOO"
    }
  }
}

resource "null_resource" "null" {
  triggers = {
    hello: "world"
  }
}
```
</details>

laurapacilio commented on website/docs/cli/config/config-file.mdx:
```diff
@@ -115,12 +115,22 @@ is available at multiple hostnames, use only one of them consistently.
+Authorization: Bearer token whenever making service requests to the hostname
+"app.terraform.io".
+
+Internationalized domain names should be converted to their [punycode equivalent](https://www.charset.org/punycode)
+with an ACE prefix when used as a variable name. For example, token credentials
+for 例えば.com should be set in a variable called `TF_TOKEN_xn--r8j3dr99h_com`.
```
```suggestion
Convert domain names to their [punycode equivalent](https://www.charset.org/punycode) and include an ACE prefix when using them as a variable name. For example, set the token credentials for 例えば.com in a variable called `TF_TOKEN_xn--r8j3dr99h_com`.
```

brandonc replied:
Thanks for your suggestions! I think something was lost in this modification, because this only applies to domain names containing non-ASCII characters, typical of domain names with accent marks or unicode characters for non-latin character sets. Should it say...

> Internationalized domain names containing non-ASCII chacters should be converted

?

laurapacilio replied:
Is "internationalized" a common term? If not, I think "Convert domain names with non-ASCII characters to...."  

Thank you for pushing back on that!

apparentlymart replied:
"[Internationalized domain name](https://en.wikipedia.org/wiki/Internationalized_domain_name)" as a whole is the particular jargon phrase used to refer to the family of specifications that this "punycode" formatting is a part of. Since we're already linking to a reference for "punycode" anyway we may not need to _also_ say "Internationalized domain name", but if we do then I'd suggest linking to a definition for it so it's clearer it's not just a regular adjective modifying "domain name".

laurapacilio reviewed (changes requested):
Some thoughts on language for you - please push back where I've materially changed the meaning :)

alisdair reviewed (commented):
Code changes look good to me! A couple of thoughts:

- It's unfortunate that this approach doesn't support addresses with dashes when using bash, as you mention in the PR description. Did you consider any changes to the variable naming to support this? For example, maybe we could encode dashes as double underscores, since it's invalid to have two periods in a row.
- Should we clarify the priority order for credentials in the documentation somewhere? I had a quick look and couldn't see if it was already documented, but maybe I missed something.

laurapacilio reviewed (commented):
@brandonc I was just reading the RFC and I was wondering if we should also add the following to this docs page:
- Information about the credentials source priority order. Especially the fact that now, the environment takes precedence over the configuration.
- Some type of warning for folks who are using credential helper --> as I understand it, there is the risk of behavior change for those folks. RFC mentioned providing them with a helpful error message (and I admit I haven't looked through the code in this PR), so if that will be present for users, I feel comfortable not including that note in the docs, as it should be a relatively small subset of cases, and the users will get direct feedback in the CLI warning them about this.
- Explain to users that if there are two environment variables whose names both evaluate to the same hostname after normalization, Terraform uses the one defined later in the operating system's environment variable table for the process. Users should avoid this if possible, right?

brandonc commented:
@alisdair I did some investigation and it so happens that hyphens can't be used at the beginning or end of a label. That means that I should be able to translate the first underscore of a series of underscores with a period and then the remaining ones can be translated to dashes.

brandonc commented:
@alisdair my instinct when I made the last comment turned out to be invalid because I was thinking of the problem backwards, but your suggestion worked out well. I implemented it as a fallback for cases when a hyphen was detected in the host and added documentation that felt straightforward.

@laurapacilio I don't think its possible to set two different env vars that evaluate to the same host, except in the case of hyphens. I added that detail as well as the other two bullet points you suggested. Your feedback is tremendous, thanks!

apparentlymart commented:
It is in principle possible to create two environment variables that differ only in whether they escape dots as underscores or not, or in using uppercase vs. lowercase letters which will get folded as part of the nameprep algorithm that the IDNA library does for us.

In practice I think it unlikely that someone would do that even accidentally because the usual conventions for variable naming and the parsing rules in common shells discourage it, so I don't think we need to spend a lot of words on it but I think worth a passing mention just in case it helps someone who is trying to debug some odd behavior and so needs that prompt to think about checking if other conflicting variables might be set.

laurapacilio commented on website/docs/cli/config/config-file.mdx:
```diff
@@ -115,12 +115,29 @@ is available at multiple hostnames, use only one of them consistently.
+For example, the value of a variable named `TF_TOKEN_app_terraform_io` will be used as a
+bearer authorization token when the CLI makes service requests to the hostname "app.terraform.io".
+
+Domain names containing non-ASCII characters should be converted to
+their [punycode equivalent](https://www.charset.org/punycode)
+with an ACE prefix when used as a variable name. For example, token credentials
+for 例えば.com must be set in a variable called `TF_TOKEN_xn--r8j3dr99h_com`.
```
Just working on flow and making sure we're using active voice :-) 

```suggestion
When using domain names as a variable name, you must convert domain names containing non-ASCII characters to their [punycode equivalent](https://www.charset.org/punycode) with an ACE prefix. For example, token credentials for 例えば.com must be set in a variable called `TF_TOKEN_xn--r8j3dr99h_com`.
```

laurapacilio commented on website/docs/cli/config/config-file.mdx:
```diff
@@ -115,12 +115,29 @@ is available at multiple hostnames, use only one of them consistently.
+with an ACE prefix when used as a variable name. For example, token credentials
+for 例えば.com must be set in a variable called `TF_TOKEN_xn--r8j3dr99h_com`.
+
+Hyphens are allowed within variable names by certain tools, such as the `env` utility,
+but are not valid POSIX variable names. Therefore, it can be tricky to
+set them with interactive tools such as bash or zsh so you may encode
+hyphens as double underscores. For example, a token for the domain name "café.fr" can
+be set by either `TF_TOKEN_xn--caf-dma_fr` or `TF_TOKEN_xn____caf__dma_fr` if that is more
+convenient. If both are defined, the variable containing hyphens is preferred.
```
I'm working on active voice and also just making things clearer by fussing with the way information is presented in the sentences. As usual - please push back on me if I've edited in a way that loses/changes the meaning :-) Thank you so much for working with me on this! I appreciate you adding in these extra details for users.

```suggestion
Some tools like the `env` utility allow hyphens in variable names, but hyphens create invalid POSIX variable names. Therefore, you can encode hyphens as double underscores when you set variables with interactive tools like Bash or Zsh. For example, you can set a token for the domain name "café.fr" as either `TF_TOKEN_xn--caf-dma_fr` or `TF_TOKEN_xn____caf__dma_fr`. If both are defined, Terraform will use the version containing hyphens.
```

laurapacilio reviewed (changes requested):
So close - i like the extra context you added! I made a few more adjustments, please push back where needed!
